### PR TITLE
titus2 1.2.1 ❄️

### PR DESCRIPTION
**Pull request by boldorider4**
titus2 1.2.1 :snowflake:

**Destination channel:** Snowflake

### Links

- [PKG-3787]
- [Upstream repository](https://github.com/animator/titus2)
- Relevant PRs:
  - [avro-python3](https://github.com/AnacondaRecipes/avro-python3-feedstock/pull/3)

### Notes:

- `py310`, `py311` skipped as per [upstream requirement](https://github.com/animator/titus2/blob/f1712196d7ad705c6598f13b24bcb1e580bc93fe/setup.py#L66)
- `avro-python3` is [specified](https://github.com/animator/titus2/blob/f1712196d7ad705c6598f13b24bcb1e580bc93fe/setup.py#L36) as a dependency although the package is [deprecated](https://pypi.org/project/avro-python3/) in favor of `avro`

[PKG-3787]: https://anaconda.atlassian.net/browse/PKG-3787?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by lorepirri** [comment]
tests fail with:

```
...
ModuleNotFoundError: No module named 'avro.io'
...
ModuleNotFoundError: No module named 'avro.datafile'
...
```

both should come from `avro-python3 v1.10.2` that we just deployed in https://github.com/AnacondaRecipes/avro-python3-feedstock/pull/2

See https://inspector.pypi.io/project/avro-python3/1.10.2

titus2 depends on `avro-python3`, which seems to be deprecated, and in turn, depends on `avro`.

titus2 metadata:
https://inspector.pypi.io/project/titus2/1.2.1/packages/29/26/c0e1237b090986b1f11277b6827905d20b1ddaaea5eb203e70abb42a9a10/titus2-1.2.1-py3-none-any.whl/titus2-1.2.1.dist-info/METADATA#line.35

**Comment by lorepirri** [comment]
I skip `testUnpackBoolean` test on `osx` because it fails for no apparent reason with:

```
======================================================================
FAIL: testUnpackBoolean (testGenpy.TestGeneratePython)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/lorepirri/miniconda3/conda-bld/titus2_1708542245206/test_tmp/test/testGenpy.py", line 2114, in testUnpackBoolean
    self.assertEqual(engine.action("".join(map(unsigned, [8]))), True)
AssertionError: False != True

----------------------------------------------------------------------
```

in case, I can investigate further but it seems to me that it depends on the package and not on the build environment. The function is supposed to convert a byte string `b'\x08'` to boolean, which is `True` on other systems but `False` on `osx`.

**Comment by lorepirri** [comment]
> tests fail with:
> 
> ```
> ...
> ModuleNotFoundError: No module named 'avro.io'
> ...
> ModuleNotFoundError: No module named 'avro.datafile'
> ...
> ```
> 
> both should come from `avro-python3 v1.10.2` that we just deployed in [AnacondaRecipes/avro-python3-feedstock#2](https://github.com/AnacondaRecipes/avro-python3-feedstock/pull/2)
> 
> See https://inspector.pypi.io/project/avro-python3/1.10.2
> 
> titus2 depends on `avro-python3`, which seems to be deprecated, and in turn, depends on `avro`.
> 
> titus2 metadata: https://inspector.pypi.io/project/titus2/1.2.1/packages/29/26/c0e1237b090986b1f11277b6827905d20b1ddaaea5eb203e70abb42a9a10/titus2-1.2.1-py3-none-any.whl/titus2-1.2.1.dist-info/METADATA#line.35

this issue was resolved by fixing `avro-python3` and until that is merged and deployed, this PR uses a staging channel, remember to remove it before merging.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    - titus.lib.prob
+    - titus.lib.stat
+    - titus.inspector
+  requires:
+    - pip
+  source_files:
+    - test
+  commands:
+    - pip check
+    - python -m unittest discover -s test -p "test*.py"
```
Comment (ViridianMelody) [inline recipe/meta.yaml]:
```suggestion
    - pip
    - pytest
  source_files:
    - test
  commands:
    - pip check
    - pytest test -o python_files=test*.py -k "not testUnpackBoolean" -v  # [osx]
    - pytest test -o python_files=test*.py -v  # [not osx]
```

Pytest can skip tests from the command line. Easier to maintain than patching test files directly.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+build:
+  skip: True  # [py>=310]
+  # missing win build for avro --> avro-python3
+  skip: True  # [win]
+  script: {{ PYTHON }} -m pip install . --no-deps --no-build-isolation -vv
+  number: 0
+
+requirements:
+  build:
+    - patch  # [osx]
```
Comment (ViridianMelody) [inline recipe/meta.yaml]:
```suggestion

build:
  skip: True  # [py>=310]
  # missing win build for avro --> avro-python3
  skip: True  # [win]
  script: {{ PYTHON }} -m pip install . --no-deps --no-build-isolation -vv
  number: 0

requirements:
```

Patch is unnecessary if using the pytest solution I mentioned.

Comment (boldorider4) [reply recipe/meta.yaml]: Agreed. Thanks for the great suggestion!

**Review by lorepirri — APPROVED** [review]
LGTM! Thank you @ViridianMelody for the fix! It's a great idea! Thanks @boldorider4 for all the help!

**Review by ViridianMelody — APPROVED** [review]
LGTM

**Pull request merged**